src/socket/socket.ts

```diff
@@ -0,0 +1,60 @@
+import { Socket } from "socket.io";
+
+import { webSocket } from "..";
+import { Product } from "../models/product.model";
+import { findProductQuery, checkoutQuery } from "../database/product.queries";
+import { logger } from "../logger/logger";
+
+const carts: Record<string, Record<string, number>> = {};
+
+export const registerSocket = (socket: Socket, next: (err?: Error) => void) => {
+    logger.info(`A user connected ${socket.id}`);
+    carts[socket.id] = {};
+    next();
+}
+
+const clearCart = async (socket: Socket) =>
+    await Promise.all(Object.keys(carts[socket.id])
+        .map((id) => manageProductInCart(socket)({ id, amount: 0 })))
+
+export const removeSocket = (socket: Socket) => async () => {
+    logger.info(`A user disconnected ${socket.id}`);
+    await clearCart(socket);
+    delete carts[socket.id];
+}
+
+export const calculateProductLimit = (id: string): number =>
+    Object.values(carts).reduce((acc, productInCart) =>
+        productInCart[id] ? acc + productInCart[id] : acc, 0);
+
+const updateProductCartAmount = (socketId: string, product: Product, amount: number) => {
+    amount === 0 ? delete carts[socketId][product._id] : carts[socketId][product._id] = amount;
+    if (product.limit !== undefined) {
+        const limit: number = product.limit - calculateProductLimit(product._id);
+        webSocket.emit('productChanges', { ...product.toObject(), limit });
+    }
+}
+
+export const manageProductInCart = (socket: Socket) => {
```

Comment:
add return type

Reply:
shloud I? its just coping the line below like:
const manageProductInCart = (socket: Socket): ({ id, amount }: { id: string, amount: number }) => void => {
    return async ({ id, amount }: { id: string, amount: number }) => { } }